### PR TITLE
os/bluestore: fix reclaim_blocks and clean up Allocator interface

### PR DESCRIPTION
**Pull request by liewegas**
This evolved into something truly ugly, and reclaim_blocks managed to misuse
it.  Fix and clean.

### CODE REVIEW

**Comment by liewegas** [comment]
@chhabaramesh

**Comment by chhabaramesh** [comment]
@liewegas , overall changes looks good. Only thing is if emblace_back could cause multiple allocations in single allocator call? are we worried about that?

**Comment by liewegas** [comment]
We can do a reserve() call in the caller for that if we want. I suspect that a small number like 2 or 4 is sufficient.

**Comment by chhabaramesh** [comment]
yes, I think that would suffice since most of allocations are single to few blocks.

File: src/os/bluestore/bluestore_types.cc [hunk 1]
```diff
@@ -21,8 +21,8 @@ void ExtentList::add_extents(int64_t start, int64_t count) {
   AllocExtent *last_extent = NULL;
   bool can_merge = false;
 
-  if (m_num_extents > 0) {
-    last_extent = &((*m_extents)[m_num_extents - 1]);
+  if (!m_extents->empty()) {
+    last_extent = &((*m_extents)[m_extents->size() - 1]);
```
Comment (ifed01) [inline src/os/bluestore/bluestore_types.cc]: replace with 
last_extent = &(m_extents->back());
?

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -1765,6 +1765,7 @@ int BlueFS::_allocate(uint8_t id, uint64_t len,
   }
 
   AllocExtentVector extents;
+  extents.reserve(4);
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: May be some comment about an arbitrary selection of 4?

File: src/test/objectstore/BitAllocator_test.cc [hunk 3]
```diff
@@ -283,19 +283,18 @@ TEST(BitAllocator, test_zone_alloc)
   bmap_test_assert(lock);
 
   int64_t blk_size = 1024;
-  AllocExtentVector extents = AllocExtentVector
-        (zone->size() / 2, AllocExtent(-1, -1));
-
+  AllocExtentVector extents;
+  cout << extents.size() << std::endl;
```
Comment (ifed01) [inline src/test/objectstore/BitAllocator_test.cc]: 3 odd lines...

File: src/test/objectstore/BitAllocator_test.cc [hunk 4]
```diff
@@ -371,36 +371,36 @@ TEST(BitAllocator, test_zone_alloc)
...
-        allocated = zone->alloc_blocks_dis(total_blocks, total_blocks / 4, 0, 0, block_list);
-        bmap_test_assert(allocated == total_blocks);
-        for (int i = 0; i < 4; i++) {
-          bmap_test_assert(extents[i].offset == (uint64_t) i * (total_blocks / 4));
-        }
-  
+      delete block_list;
+      extents.clear();
+      block_list = new ExtentList(&extents, blk_size, total_blocks / 4 * blk_size);
+      block_list->reset();
```
Comment (ifed01) [inline src/test/objectstore/BitAllocator_test.cc]: This duplicates extents.clear() call above. Here and throughout.

**Pull request merged**